src/katgpucbf/xbgpu/engine.py

```diff
@@ -254,25 +266,180 @@ async def sender_loop(self) -> None:
+            tx_item.add_marker(self._proc_command_queue)
+            self._tx_item_queue.put_nowait(tx_item)
+
+            # Finish with the rx_item
+            rx_item.add_marker(self._proc_command_queue)
+            self.engine.free_rx_item(rx_item)
+        # When the stream is closed, if the sender loop is waiting for a tx item,
+        # it will never exit. Upon receiving this NoneType, the sender_loop can
+        # stop waiting and exit.
+        logger.debug("gpu_proc_loop completed")
+        self._tx_item_queue.put_nowait(None)
+
+    async def sender_loop(self) -> None:  # noqa: D102
+        # NOTE: This function passes the entire downloaded data to
+        # chunk.send, which then takes care of directing data to each beam's
+        # output destination.
+
+        while True:
+            # Get a populated TxQueueItem from the tx_item_queue
+            item = await self._tx_item_queue.get()
+            if item is None:
+                break
+            # The CPU doesn't need to wait, but the GPU does to ensure it
+            # won't start the download before computation is done.
+            item.enqueue_wait_for_events(self._download_command_queue)
+
+            # Get a free Chunk from the send_stream
+            chunk = await self.send_stream.get_free_chunk()
+
+            # Kick off any data downloads from the GPU
+            item.buffer_device.get_async(self._download_command_queue, chunk.data)
+            item.saturated.get_async(self._download_command_queue, chunk.saturated)
+
+            # Wait for the transfer
+            event = self._download_command_queue.enqueue_marker()
+            await katsdpsigproc.resource.async_wait_for_events([event])
+
+            # Set the Chunk timestamp
+            chunk.timestamp = item.timestamp
+            if self.send_stream.tx_enabled:
+                # TODO: Update beng-clip-cnt sensor
+                # But fgpu does it all in chunk.send by passing it
+                # - Send stream(s),
+                # - Frames (int, to send)
+                # - time_converter, to get `end_time` for setting clip-cnt sensor
+                # - engine.sensors, to obtain the clip-cnt sensor
+                # - output_name, because the Chunk didn't have access to it
+                #   (to find sensor name).
+                pass
+            self.send_stream.send_chunk(chunk, self.engine.time_converter, self.engine.sensors)
+            self._tx_free_item_queue.put_nowait(item)
+
+        await self.send_stream.send_stop_heap()
+        logger.debug("sender_loop completed")
+
+    def capture_enable(self, *, stream_id: int, enable: bool = True) -> None:  # noqa: D102
+        self.send_stream.enable_substream(stream_id=stream_id, enable=enable)
+
+
 class XPipeline(Pipeline):
     """Processing pipeline for a single baseline-correlation-products stream."""
 
     def __init__(
         self,
-        output: XOutput,
+        outputs: Sequence[XOutput],
```

Comment:
Since XPipeline can only work on a single output, I'd suggest reverting this change, and passing `[output]` to the base class constructor.

A lot of other changes in the constructor (where you changed `output` to `self.output`) can then also be reverted.